// request-json/main.js
// Generated by CoffeeScript 1.4.0
(function() {
  var fs, parseBody, request, url;

  request = require("request");

  fs = require("fs");

  url = require("url");

  parseBody = function(error, response, body, callback) {
    var parsed;
    if (typeof body === "string" && body !== "") {
      try {
        parsed = JSON.parse(body);
      } catch (err) {
        if (error == null) {
          error = err;
        }
        parsed = body;
      }
    } else {
      parsed = body;
    }
    return callback(error, response, parsed);
  };

  exports.JsonClient = (function() {

    JsonClient.prototype.agent = "request-json/1.0";

    function JsonClient(host) {
      this.host = host;
    }

    JsonClient.prototype.setBasicAuth = function(username, password) {
      var basicCredentials, credentials;
      credentials = "" + username + ":" + password;
      basicCredentials = new Buffer(credentials).toString('base64');
      return this.auth = "Basic " + basicCredentials;
    };

    JsonClient.prototype.setToken = function(token) {
      return this.token = token;
    };

    JsonClient.prototype.get = function(path, callback, parse) {
      var header;
      if (parse == null) {
        parse = true;
      }
      header = {
        accept: 'application/json'
      };
      if (this.auth != null) {
        header["authorization"] = this.auth;
      }
      if (this.agent != null) {
        header["user-agent"] = this.agent;
      }
      if (this.token != null) {
        header["x-auth-token"] = this.token;
      }
      return request({
        method: 'GET',
        headers: header,
        uri: url.resolve(this.host, path)
      }, function(error, response, body) {
        if (parse) {
          return parseBody(error, response, body, callback);
        } else {
          return callback(error, response, body);
        }
      });
    };

    JsonClient.prototype.post = function(path, json, callback, parse) {
      if (parse == null) {
        parse = true;
      }
      return request({
        method: "POST",
        uri: url.resolve(this.host, path),
        json: json,
        headers: {
          authorization: this.auth,
          "user-agent": this.agent,
          'x-auth-token': this.token
        }
      }, function(error, response, body) {
        if (parse) {
          return parseBody(error, response, body, callback);
        } else {
          return callback(error, response, body);
        }
      });
    };

    JsonClient.prototype.put = function(path, json, callback) {
      return request({
        method: "PUT",
        uri: url.resolve(this.host, path),
        json: json,
        headers: {
          authorization: this.auth,
          "user-agent": this.agent,
          'x-auth-token': this.token
        }
      }, function(error, response, body) {
        return parseBody(error, response, body, callback);
      });
    };

    JsonClient.prototype.del = function(path, callback) {
      return request({
        method: "DELETE",
        uri: url.resolve(this.host, path),
        headers: {
          authorization: this.auth,
          'x-auth-token': this.token
        }
      }, function(error, response, body) {
        return parseBody(error, response, body, callback);
      });
    };

    JsonClient.prototype.sendFile = function(path, filePath, data, callback) {
      var att, form, req;
      if (typeof data === "function") {
        callback = data;
      }
      req = this.post(path, null, callback, false);
      form = req.form();
      if (typeof data !== "function") {
        for (att in data) {
          form.append(att, data[att]);
        }
      }
      return form.append('file', fs.createReadStream(filePath));
    };

    JsonClient.prototype.saveFile = function(path, filePath, callback) {
      var stream;
      stream = this.get(path, callback, false);
      return stream.pipe(fs.createWriteStream(filePath));
    };

    return JsonClient;

  })();

}).call(this);
